clang/lib/Sema/Sema.cpp

```diff
@@ -602,68 +602,133 @@ ExprResult Sema::ImpCastExprToType(Expr *E, QualType Ty,
+    Optional<unsigned> DiagID;
+    if (FromIsCap && !ResultIsCap && ExprTy->isPointerType()) {
+      // T* __capability -> T* is never allowed implicitly.
+      DiagID = diag::err_typecheck_convert_cap_to_ptr;
+    } else if (ResultIsCap && !FromIsCap) {
+      // ImpCastPointerToCHERICapability can create new implicit
+      // CK_FunctionToPointerDecay/CK_ArrayToPointerDecay so we need to avoid
+      // infinite recursion here
+      if (!ImpCastPointerToCHERICapability(ExprTy, TypeTy, E, true))
+        return ExprError();               // diagnostics already emitted
+      DiagID = None;                      // got a permitted explicit conversion
+      Kind = CK_PointerToCHERICapability; // update cast-kind for the new cast
+    }
+    if (DiagID) {
+      // Note: ImpCastPointerToCHERICapability may have changed E, so we must
+      // use E->getType() instead of ExprTy
+      auto Fixit = FixItHint::CreateInsertion(
+          E->getBeginLoc(), "(__cheri_" +
+                                std::string(FromIsCap ? "from" : "to") +
+                                "cap " + TypeTy.getAsString() + ")");
+      return ExprError(Diag(E->getBeginLoc(), *DiagID)
+                       << E->getType() << TypeTy << false << Fixit);
     }
   }
 
   return ImplicitCastExpr::Create(Context, Ty, Kind, E, BasePath, VK);
 }
 
-bool Sema::ImpCastPointerToCHERICapability(QualType FromTy, QualType ToTy, Expr *&From, bool Diagnose) {
-
-  assert(FromTy->isPointerType() && ToTy->isPointerType() && ToTy->getAs<PointerType>()->isCHERICapability() &&
-         "Both argument types should be PointerType");
-
+bool Sema::ImpCastPointerToCHERICapability(QualType FromTy, QualType ToTy,
+                                           Expr *&From, bool Diagnose) {
+  assert(ToTy->isPointerType() && "Target types should be PointerType");
+  assert(ToTy->getAs<PointerType>()->isCHERICapability() &&
+         "Target type must be a capability pointer");
+  if (From->isNullPointerConstant(Context, Expr::NPC_ValueDependentIsNotNull)) {
+    return true; // NULL constants are always fine
+  }
   bool StrLit = dyn_cast<StringLiteral>(From->IgnoreImpCasts()) != nullptr;
-  bool NullConstant = From->isNullPointerConstant(Context, Expr::NPC_ValueDependentIsNotNull);
   bool AddrOf = false;
   bool Decayed = false;
   bool CompatTypes = false;
 
-  // address-of (&)
-  if (UnaryOperator *UnOp = dyn_cast<UnaryOperator>(From)) {
-    if (UnOp->getOpcode() == UO_AddrOf) {
-      AddrOf = true;
-      CompatTypes = CheckCHERIAssignCompatible(ToTy, FromTy, From);
-    }
-  }
   // Function-to-pointer and array-to-pointer decay
-  if (ImplicitCastExpr* Imp = dyn_cast<ImplicitCastExpr>(From)) {
+  ExprResult DecayedExpr = DefaultFunctionArrayConversion(From, Diagnose);
+  if (DecayedExpr.isUsable() && !DecayedExpr.isInvalid()) {
+    From = DecayedExpr.get();
+    FromTy = From->getType();
+  }
+
+  // FIXME: clean up control-flow in this function
```

Comment:
Could be done very easily by replacing `goto err;` with `return SomeLambda(...);` (or `SomeLambda(..); return false`) where `SomeLambda` is defined just above with the body of the current `err` label.